File: Seminar/seminar_5/src/main/java/ubb/scs/map/ir/seminar/studentsmanagement/repository/InMemoryRepository.java
package ubb.scs.map.ir.seminar.studentsmanagement.repository;

import ubb.scs.map.ir.seminar.studentsmanagement.domain.Entity;
import ubb.scs.map.ir.seminar.studentsmanagement.domain.validators.Validator;

import java.util.HashMap;
import java.util.Map;

public class InMemoryRepository<ID, E extends Entity<ID>> implements CrudRepository<ID,E> {
    private Map<ID, E> entities;
    private Validator<E> validator;

    public InMemoryRepository(Validator<E> validator) {
        this.validator = validator;
        entities = new HashMap<ID, E>();
    }

    @Override
    public E findOne(ID id) {
        return null;
    }


    @Override
    public Iterable<E> findAll() {
        return entities.values();
    }

    @Override
    public E save(E entity) {
        if(entity == null) throw new IllegalArgumentException("entity must be not null");
        if(entities.containsKey(entity.getId())) return entity;
        validator.validate(entity);
        entities.put(entity.getId(), entity);
        return null;
    }

    @Override
    public E delete(ID id) {
        return null;
    }

    @Override
    public E update(E entity) {
        return null;
    }
}

File: Laborator/Lab 4/Laborator 9/Laborator 9/Repository/InMemoryRepository.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using Laborator_9.Repository;
using Laborator_9.Domain;
using Laborator_9.Domain.Validators;

namespace Laborator_9.Repository
{
    class InMemoryRepository<ID, E> : IRepository<ID, E> where E : Entity<ID>
    {
        protected IValidator<E> validator;
        protected Dictionary<ID, E> elements = new Dictionary<ID, E>();

        public InMemoryRepository(IValidator<E> validator)
        {
            this.validator = validator;
        }

        public virtual E Delete(ID id)
        {
            if (id == null)
                throw new ArgumentNullException("ID must not be null");
            if (!elements.ContainsKey(id))
                return default(E);
            E value = elements[id];
            elements.Remove(id);
            return value;
        }

        public IEnumerable<E> FindAll()
        {
            return elements.Values;
        }

        public E FindOne(ID id)
        {
            if (id == null)
                throw new ArgumentNullException("ID must not be null");
            return elements[id];
        }

        public virtual E Save(E entity)
        {
            if (entity == null)
                throw new ArgumentNullException("entity must not be null");
            if (elements.ContainsKey(entity.ID))
                return entity;

            validator.Validate(entity);
            elements[entity.ID] = entity;
            return default(E);
        }

        public virtual E Update(E entity)
        {
            if (entity == null)
                throw new ArgumentNullException("entity must not be null");
            if (!elements.ContainsKey(entity.ID))
                return entity;
            
            validator.Validate(entity);
            elements[entity.ID] = entity;
            return default(E);
        }
    }
}
